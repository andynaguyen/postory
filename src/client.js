const request = (carrier, trackingNumber) => {
  return fetch(`${process.env.ENDPOINT}/track?carrier=${carrier}&trackingNumber=${trackingNumber}`).then((res) =>
    res.json(),
  );
};

export default { request };
