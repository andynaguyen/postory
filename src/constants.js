export const ImgFileByCarrier = {
  fedex: 'images/fedex.webp',
  dhl_express: 'images/dhl.webp',
  ups: 'images/ups.webp',
  usps: 'images/usps.png',
};
