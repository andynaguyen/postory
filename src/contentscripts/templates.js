import { html } from 'lit-html';

export const progressBar = () =>
  html`
    <progress class="postory-progress is-small is-primary" max="100">15%</progress>
  `;

export const trackingAlertContainer = ({ carrier, trackingNumber }) =>
  html`
    <iframe
      src="${chrome.runtime.getURL('iframe.html')}?carrier=${carrier}&trackingNumber=${trackingNumber}"
      id="postory-tracking-alert-iframe"
    ></iframe>
    <div id="postory-progress-bar"></div>
  `;

export const trackingAlert = ({ trackingNumber, carrierImgSrc, carrierImgAlt, onDismiss, onConfirm }) =>
  html`
    <div class="card animated fadeIn">
      <header class="card-header">
        <p class="card-header-title">
          Track shipment using Postory?
        </p>
      </header>
      <div class="card-content">
        <div class="media">
          <div class="media-left">
            <figure class="image is-48x48">
              <img src=${carrierImgSrc} alt=${carrierImgAlt} />
            </figure>
          </div>
          <div class="media-content">
            <input id="display-name-input" class="input" type="text" placeholder="Display name (optional)" />
          </div>
        </div>
      </div>
      <footer class="card-footer">
        <a href="#" class="card-footer-item" @click=${onDismiss}>Not Now</a>
        <a href="#" class="card-footer-item" @click=${onConfirm}>OK</a>
      </footer>
    </div>
  `;
