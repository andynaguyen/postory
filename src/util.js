import { DateTime } from 'luxon';

const formatDateTime = (dateTime) => dateTime.toLocaleString(DateTime.DATETIME_MED);

export default {
  formatDateTime,
};
