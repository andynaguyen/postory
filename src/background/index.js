import TrackingInfoStorage from 'src/storage';
import { checkForTrackingInfoUpdate } from 'src/background/check';
import { createStatusNotification } from 'src/notification';

const storage = new TrackingInfoStorage();
const interval = 1000 * 60 * 60 * 2; // 2 hours

const isTerminalStatus = (trackingInfo) => {
  const status = trackingInfo.tracking_status.status;
  return status === 'DELIVERED' || status === 'RETURNED' || status === 'FAILURE';
};

setInterval(() => {
  storage.list((trackingInfos) => {
    trackingInfos.forEach((trackingInfo) => {
      if (isTerminalStatus(trackingInfo)) {
        checkForTrackingInfoUpdate(trackingInfo).then((result) => {
          if (result.isUpdated) {
            createStatusNotification(result.trackingInfo);
            storage.put(result.trackingInfo);
          }
        });
      }
    });
  });
}, interval);
