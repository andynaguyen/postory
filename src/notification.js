import { ImgFileByCarrier } from 'src/constants';

export const createStatusNotification = (trackingInfo) => {
  chrome.notifications.create(undefined, {
    type: 'basic',
    iconUrl: ImgFileByCarrier[trackingInfo.carrier],
    title: `Postory - ${trackingInfo.tracking_number}`,
    message: `Your package has changed status: ${trackingInfo.tracking_status.status}`,
  });
};
