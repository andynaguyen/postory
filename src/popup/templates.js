import { DateTime } from 'luxon';
import { html } from 'lit-html';
import { ImgFileByCarrier } from 'src/constants';
import { openCarrierTab } from 'src/tab';
import util from 'src/util';

const BadgeInfoByTrackingStatus = {
  DELIVERED: {
    text: 'Delivered',
    className: 'is-primary',
  },
  PRE_TRANSIT: {
    text: 'Pre-transit',
    className: 'is-info',
  },
  TRANSIT: {
    text: 'Transit',
    className: 'is-info',
  },
  RETURNED: {
    text: 'Returned',
    className: 'is-danger',
  },
  FAILURE: {
    text: 'Failure',
    className: 'is-danger',
  },
  UNKNOWN: {
    text: 'Unknown',
    className: 'is-light',
  },
};

const trackingInfoStatusBadge = (trackingInfoStatus) => {
  const { text, className } = BadgeInfoByTrackingStatus[trackingInfoStatus] || BadgeInfoByTrackingStatus.UNKNOWN;
  return html`
    <div class="tags has-addons">
      <span class="tag">Status</span>
      <span class="tag ${className}">${text}</span>
    </div>
  `;
};

const trackingInfoCardContent = (trackingInfo) =>
  html`
    <div class="card-content">
      <div class="media">
        <div class="media-left">
          <figure class="image is-48x48">
            <img src="${ImgFileByCarrier[trackingInfo.carrier]}" alt="${trackingInfo.carrier}" />
          </figure>
        </div>
        <div class="media-content tracking-info-media-content">
          <div class="tags has-addons">
            <span class="tag">ETA</span>
            <span class="tag is-info">${util.formatDateTime(DateTime.fromISO(trackingInfo.eta))}</span>
          </div>
          ${trackingInfoStatusBadge(trackingInfo.tracking_status.status)}
        </div>
      </div>
    </div>
  `;

const trackingInfoListItem = (trackingInfo, onDeleteTrackingInfo) => {
  const id = `${trackingInfo.carrier}#${trackingInfo.tracking_number}`;
  return html`
    <li id="${id}" class="container tracking-info">
      <div class="card tracking-info-card">
        <header class="card-header">
          <p class="card-header-title with-pointer" @click=${() => openCarrierTab(trackingInfo)}>
            ${trackingInfo.displayName || trackingInfo.tracking_number}
          </p>
          <div class="card-header-icon no-pointer">
            <button class="card-header-icon delete" aria-label="delete" @click=${() => onDeleteTrackingInfo(id)} />
          </div>
        </header>
        ${trackingInfoCardContent(trackingInfo)}
      </div>
    </li>
  `;
};

export const trackingInfoList = (trackingInfos, onDeleteTrackingInfo) =>
  html`
    <ul id="tracking-info-list">
      ${trackingInfos.length
        ? trackingInfos.map((trackingInfo) => trackingInfoListItem(trackingInfo, onDeleteTrackingInfo))
        : html`
            <i>Nothing added to Postory yet. Navigate to a tracking page to track the shipment!</i>
          `}
    </ul>
  `;
