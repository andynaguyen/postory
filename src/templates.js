import { DateTime } from 'luxon';
import { html } from 'lit-html';
import { ImgFileByCarrier } from 'src/constants';
import util from 'src/util';

const trackingInfoListItem = (trackingInfo, onDeleteTrackingInfo) => {
  const id = `${trackingInfo.carrier}#${trackingInfo.tracking_number}`;
  return html`
    <li id="${id}" class="container tracking-info">
      <div class="card">
        <header class="card-header">
          <p class="card-header-title">${trackingInfo.displayName || trackingInfo.tracking_number}</p>
          <div class="card-header-icon no-cursor">
            <button class="card-header-icon delete" aria-label="delete" @click=${() => onDeleteTrackingInfo(id)} />
          </div>
        </header>
        <div class="card-content">
          <div class="media">
            <div class="media-left">
              <figure class="image is-48x48">
                <img src="${ImgFileByCarrier[trackingInfo.carrier]}" alt="${trackingInfo.carrier}" />
              </figure>
            </div>
            <div class="media-content tracking-info-media-content">
              <p>ETA: ${util.formatDateTime(DateTime.fromISO(trackingInfo.eta))}</p>
              <p>STATUS: ${trackingInfo.tracking_status.status}</p>
            </div>
          </div>
        </div>
      </div>
    </li>
  `;
};

export const trackingInfoList = (trackingInfos, onDeleteTrackingInfo) =>
  html`
    <ul id="tracking-info-list">
      ${trackingInfos.map((trackingInfo) => trackingInfoListItem(trackingInfo, onDeleteTrackingInfo))}
    </ul>
  `;

export const progressBar = () =>
  html`
    <progress class="postory-progress is-small is-primary" max="100">15%</progress>
  `;

export const trackingAlertContainer = ({ carrier, trackingNumber }) =>
  html`
    <iframe
      src="${chrome.runtime.getURL('iframe.html')}?carrier=${carrier}&trackingNumber=${trackingNumber}"
      id="postory-tracking-alert-iframe"
    ></iframe>
    <div id="postory-progress-bar"></div>
  `;

export const trackingAlert = ({ trackingNumber, carrierImgSrc, carrierImgAlt, onDismiss, onConfirm }) =>
  html`
    <div class="card animated fadeIn">
      <header class="card-header">
        <p class="card-header-title">
          Track shipment using Postory?
        </p>
      </header>
      <div class="card-content">
        <div class="media">
          <div class="media-left">
            <figure class="image is-48x48">
              <img src=${carrierImgSrc} alt=${carrierImgAlt} />
            </figure>
          </div>
          <div class="media-content">
            <input id="display-name-input" class="input" type="text" placeholder="Display name (optional)" />
          </div>
        </div>
      </div>
      <footer class="card-footer">
        <a href="#" class="card-footer-item" @click=${onDismiss}>Not Now</a>
        <a href="#" class="card-footer-item" @click=${onConfirm}>OK</a>
      </footer>
    </div>
  `;
