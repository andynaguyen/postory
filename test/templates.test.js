import { TrackingInfoList } from 'src/templates';
import util from 'src/util';

describe('TrackingInfoList', () => {
  it('should generate empty ul given empty list', () => {
    const template = TrackingInfoList([]);
    expect(template).toMatchSnapshot();
  });

  it("should generate li's given tracking info list", () => {
    util.formatDateTime = jest
      .fn()
      .mockReturnValueOnce('Oct 18, 2019, 12:09 PM')
      .mockReturnValueOnce('Oct 15, 2019, 3:00 AM');
    const template = TrackingInfoList([
      {
        carrier: 'fedex',
        eta: '2019-10-18T19:09:20-0000',
        tracking_number: '12345',
        tracking_status: {
          status: 'DELIVERED',
        },
      },
      {
        carrier: 'ups',
        eta: '2019-10-15T10:00:00-0000',
        tracking_number: 'ABCDEF',
        tracking_status: {
          status: 'TRANSIT',
        },
      },
    ]);
    expect(template).toMatchSnapshot();
  });
});
