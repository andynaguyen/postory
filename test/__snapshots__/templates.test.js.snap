// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TrackingInfoList should generate empty ul given empty list 1`] = `
"
    <ul id=\\"tracking-info-list\\">
      
    </ul>
  "
`;

exports[`TrackingInfoList should generate li's given tracking info list 1`] = `
"
    <ul id=\\"tracking-info-list\\">
      
    <li id=\\"fedex#12345\\" class=\\"container tracking-info\\" >
      <div class=\\"card\\">
        <header class=\\"card-header\\">
          <p class=\\"card-header-title\\">12345</p>
          <div class=\\"card-header-icon no-cursor\\">
            <button class=\\"card-header-icon delete aria-label=\\"delete\\"\\"/>
          </div>
        </header>
        <div class=\\"card-content\\">
          <div class=\\"media\\">
            <div class=\\"media-left\\">
              <figure class=\\"image is-48x48\\">
                <img src=\\"images/fedex.webp\\" alt=\\"fedex\\"/>
              </figure>
            </div>
            <div class=\\"media-content tracking-info-media-content\\" >
              <p>ETA: Oct 18, 2019, 7:09 PM</p>
              <p>STATUS: DELIVERED</p>
            </div>
          </div>
        </div>
      </div>
    </li>
  
    <li id=\\"ups#ABCDEF\\" class=\\"container tracking-info\\" >
      <div class=\\"card\\">
        <header class=\\"card-header\\">
          <p class=\\"card-header-title\\">ABCDEF</p>
          <div class=\\"card-header-icon no-cursor\\">
            <button class=\\"card-header-icon delete aria-label=\\"delete\\"\\"/>
          </div>
        </header>
        <div class=\\"card-content\\">
          <div class=\\"media\\">
            <div class=\\"media-left\\">
              <figure class=\\"image is-48x48\\">
                <img src=\\"images/ups.webp\\" alt=\\"ups\\"/>
              </figure>
            </div>
            <div class=\\"media-content tracking-info-media-content\\" >
              <p>ETA: Oct 15, 2019, 10:00 AM</p>
              <p>STATUS: TRANSIT</p>
            </div>
          </div>
        </div>
      </div>
    </li>
  
    </ul>
  "
`;
